tests/test_client/tests.py

```diff
@@ -856,6 +856,36 @@ def test_external_redirect(self):
             response, "https://www.djangoproject.com/", fetch_redirect_response=False
         )
 
+    def test_external_redirect_http_host(self):
+        response_1 = HttpResponseRedirect(redirect_to="https://www.djangoproject.com")
+        response_2 = HttpResponse()
+
+        with mock.patch("django.test.Client.request") as m:
+            m.side_effect = [response_1, response_2]
+            client = self.client_class()
+            client.get("/django_project_redirect/", follow=True)
+
+        call_1_params = {
+            "PATH_INFO": "/django_project_redirect/",
+            "REQUEST_METHOD": "GET",
+            "SERVER_PORT": "80",
+            "wsgi.url_scheme": "http",
+            "QUERY_STRING": "",
+        }
+        call_2_params = {
+            "PATH_INFO": "/",
+            "REQUEST_METHOD": "GET",
+            "SERVER_PORT": "80",
+            "wsgi.url_scheme": "https",
+            "QUERY_STRING": "",
+            "SERVER_NAME": "www.djangoproject.com",
+            "HTTP_HOST": "www.djangoproject.com",
+        }
+        calls = m.mock_calls
+
+        self.assertEqual(calls[0], mock.call(**call_1_params))
+        self.assertEqual(calls[1], mock.call(**call_2_params))
```

Comment:
There is a lot of mocking in this test :disappointed: Is it not testable with normal flow? If not, there is probably nothing to fix. Have you tried adding a test for the scenario described in the ticket?

Reply:
Thanks, @felixxm. One challenge with the ticket scenario is that it's talking about a redirect to a different hostname. Mocking here lets me have a redirect to a different hostname, and verify that the test client is setting HTTP_HOST for the second request. Is there a way to test requests to two different hostnames without mocking?

Reply:
We can totally test this without mocking 👍
See PR with failing test:  https://github.com/django/django/pull/17430

With Carlton's patch, the test passes.

Reply:
Great, thanks @shangxiao. I'll close this PR and you can use yours.

Reply:
Oh my intention was to demonstrate a test alternative and you could use it as a guide... (If we agree testing that way was beneficial).  I mean, please feel free to continue, you deserve the credit for taking the initiative on this👍

Reply:
OK, either way is fine with me. I'll update this PR based on your PR then.